kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterFactory.java

```diff
@@ -14,34 +14,60 @@
+ * relying on synchronization within a filter factory implementation.</p>
+ *
  * @param <C> the type of configuration used to create the {@code Filter}. Use {@link Void} if the {@code Filter} is not configurable.
  * @param <I> The type of the initialization data
  */
 public interface FilterFactory<C, I> {
 
     /**
-     * Initializes the factory with the specified configuration.
-     * This method is guaranteed to be called at most once for each filter configuration and before any call to
+     * <p>Initializes the factory with the specified configuration.</p>
+     *
+     * <p>This method is guaranteed to be called at most once for each filter configuration and before any call to
      * {@link #createFilter(FilterFactoryContext, Object)}.
      * This method may provide extra semantic validation of the config,
-     * and returns some object (which may be the config, or some other object) which will be passed to {@link #createFilter(FilterFactoryContext, Object)}
+     * and returns some object (which may be the config, or some other object) which will be passed to {@link #createFilter(FilterFactoryContext, Object)}.
      *
      * @param context context
      * @param config configuration
+     * @return A configuration state object, specific to the given {@code config}, which will be passed to the other methods of this interface.
      * @throws PluginConfigurationException when the configuration is invalid
      */
     I initialize(FilterFactoryContext context, C config) throws PluginConfigurationException;
 
     /**
      * Creates an instance of the Filter.
      *
+     * <p>This can be called on a different thread from {@link #initialize(FilterFactoryContext, Object)} and {@link #close(I)}.
+     * Implementors should either use the {@code initializationData} to pass state from {@link #initialize(FilterFactoryContext, Object)}
+     * or use appropriate synchronization.</p>
+     *
      * @param context The runtime context for the filter's creation.
-     * @param initializationData The result of the call to {@link #initialize(FilterFactoryContext, Object)}
+     * @param initializationData The initialization data that was returned from {@link #initialize(FilterFactoryContext, Object)}.
      * @return the Filter instance.
      */
     @NonNull
     Filter createFilter(FilterFactoryContext context, I initializationData);
 
+    /**
+     * Called by the runtime to release any resources associated with the given {@code initializationData}.
+     * This is guaranteed to eventually be called for each successful call to {@link #initialize(FilterFactoryContext, Object)}.
+     * Once this method has been called {@link #createFilter(FilterFactoryContext, Object)} won't be called again.
+     * @param initializationData The initialization data that was returned from {@link #initialize(FilterFactoryContext, Object)}.
+     */
+    default void close(I initializationData) {
```

Comment:
I initially having this interface extend `AutoCloseable`, but I realised that that was an ugly fit for how we're passing per-config state between `initialize` and `createFilter` using an explicit state object. So I've added that as a parameter to `close()`. So not `AutoCloseable` any more, but it avoids `FilterFactories` themselves needing to keep references to resources they create.